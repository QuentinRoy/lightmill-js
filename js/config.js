define({
    SERVER_ADDRESS: 'http://0.0.0.0:5000',
    TARGET_EXPERIMENT: 'nonmode'
});
