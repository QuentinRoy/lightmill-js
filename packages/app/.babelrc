{
  "presets": [["env", {
    "targets": { "browsers": ">1%" },
    "modules": false,
    "loose": true
  }]],
  "plugins": ["external-helpers"]
}
