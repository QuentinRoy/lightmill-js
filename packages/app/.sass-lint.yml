rules:
  property-sort-order:
    - 1
    - order: 'concentric'
