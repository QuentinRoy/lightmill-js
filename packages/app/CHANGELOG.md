# Change Log

All notable changes to this project will be documented in this file.
See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="2.0.0-1"></a>
# 2.0.0-1 (2017-06-25)



<a name="2.0.0-0"></a>
# 2.0.0-0 (2017-06-25)


### Features

* **lightmill-app:** Statically export views. ([aa071b3](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-app/commit/aa071b3))




<a name="2.0.0-0"></a>
# 2.0.0-0 (2017-06-25)


### Features

* **lightmill-app:** Statically export views. ([aa071b3](https://github.com/QuentinRoy/lightmill-js/commit/aa071b3))
