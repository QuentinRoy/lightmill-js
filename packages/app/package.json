{
  "name": "@lightmill/app",
  "version": "2.0.0",
  "description": "Base application for lightmill-runner.",
  "main": "dist/lightmill-app.js",
  "files": [
    "dist"
  ],
  "devDependencies": {
    "babel-core": "6.26.3",
    "babel-plugin-external-helpers": "6.22.0",
    "babel-preset-env": "1.7.0",
    "cross-env": "5.1.6",
    "eslint": "4.19.1",
    "eslint-config-airbnb-base": "12.1.0",
    "eslint-plugin-import": "2.12.0",
    "jest-transform-pug": "0.1.0",
    "rollup": "0.59.3",
    "rollup-plugin-babel": "3.0.4",
    "rollup-plugin-commonjs": "9.1.3",
    "rollup-plugin-node-resolve": "3.3.0",
    "rollup-plugin-pug": "0.1.6",
    "rollup-plugin-sass": "0.7.0",
    "simulant": "0.2.2"
  },
  "dependencies": {
    "pug-runtime": "^2.0.3"
  },
  "scripts": {
    "test": "cross-env NODE_ENV=test jest",
    "coverage": "yarn test --coverage",
    "build": "cross-env NODE_ENV=production rollup --config rollup.config.js",
    "prepublish": "npm run build"
  },
  "keywords": [
    "experiment",
    "lightmill",
    "application"
  ],
  "author": "Quentin Roy <quentin@quentinroy.fr>",
  "license": "MIT",
  "ava": {
    "require": [
      "babel-register"
    ],
    "babel": "inherit"
  },
  "repository": "https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-app",
  "jest": {
    "collectCoverageFrom": [
      "src/**/*.js"
    ],
    "transform": {
      "^.+\\.jsx?$": "babel-jest",
      "\\.(pug)$": "jest-transform-pug"
    },
    "moduleNameMapper": {
      "\\.(css|styl|less|sass|scss)$": "<rootDir>/../../test-tools/jest-stub-styles.js"
    }
  }
}
