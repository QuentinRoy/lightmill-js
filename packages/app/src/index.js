import end from './end';
import initBlock from './block-init';
import crash from './crash';
import wait from './wait';

/**
 * @typedef {Object} BlockInfo
 * @property {int} blockInfo.number The number of the block.
 * @property {int} blockInfo.measuredBlockNum The number of the block amongst
 * non practice blocks.
 * @property {boolean} blockInfo.practice If it is a practice block.
 * @property {Array<{factor:{name}, name}>} blockInfo.factorValues The list of
 * factor values.
 * @property {boolean} blockInfo.subjectiveAssessment If subjective assessment should be
 * asked for.
 */

/**
 * Standard Experimentation App Base class for lightmill-client providing wait, crash, end and
 * initBlock views. This is an abstract class that does not provide a `runTrial` handle
 * (required by xp-client). It is meant to be inherited (or monkey patched).
 * The app is highly reconfigurable and any method can be safely replaced at any time.
 * @constructor
 * @abstract
 */
class XpAppBase {
  /**
   * @param  {HTMLElement} node The node where to user for the app.
   */
  constructor(node) {
    /**
     * The node where lives the app (read-only).
     * @type {HTMLElement}
     */
    Object.defineProperty(this, 'node', { value: node });
  }

  /**
   * Must be overwritten by subclasses (or dynamically replaced).
   * @param {Object} config the trial configuration
   * @return {Promise<Object>} resolved with the trial results.
   * @abstract
   */
  runTrial() {
    throw new Error('runTrial not implemented');
  }

  /**
   * Show the wait view.
   * @returns {undefined}
   */
  wait() {
    wait(this.node);
  }

  /**
   * By default start waiting when the app starts.
   * @returns {undefined}
   */
  start() {
    this.wait();
  }

  /**
   * Show the crash view.
   * @param  {String} message The error message.
   * @param  {Error} [error]  The error that has been raised.
   * @param  {Object} [run]   The current run.
   * @returns {undefined}
   */
  crash(message, error, run) {
    crash(this.node, message, error, run);
  }

  /**
   * Show the block initialization view then show the wait view.
   * @param {BlockInfo} blockInfo Information about the block.
   * @return {Promise} Resolved when the user click/tap on the view.
   */
  initBlock(blockInfo) {
    return initBlock(this.node, blockInfo).then(res => {
      this.wait();
      return res;
    });
  }

  /**
   * Show the end view.
   * @returns {undefined}
   */
  end() {
    end(this.node);
  }
}

// Set up static the view handlers as static property of XpAppBase for
// cjs.
Object.assign(XpAppBase, { end, initBlock, crash, wait });

export { end, initBlock, crash, wait };
