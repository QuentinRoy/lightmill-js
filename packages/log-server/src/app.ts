/* eslint-disable @typescript-eslint/no-namespace */
import express from 'express';
import { zodiosContext } from '@zodios/express';
import { LogFilter, Store } from './store.js';
import session from 'cookie-session';
import dotenv from 'dotenv';
import { SqliteError } from 'better-sqlite3';
import { NextFunction, Request, RequestHandler, Response } from 'express';
import { api } from './api.js';
import { csvExportStream, jsonExportStream } from './export.js';
import { pipeline } from 'stream/promises';
import log from 'loglevel';
import z from 'zod';
import cuid from 'cuid';

dotenv.config();

const ctx = zodiosContext(
  z.object({
    session: z.union([
      z.null(),
      z
        .object({
          role: z.union([z.literal('admin'), z.literal('participant')]),
          runs: z.array(
            z.object({
              runId: z.string(),
              experimentId: z.string(),
            })
          ),
        })
        .strict(),
    ]),
  })
);

type CreateAppParams = {
  store: Store;
  secret: string;
  adminPassword?: string;
};
export function createLogServer({
  store,
  secret,
}: CreateAppParams): RequestHandler {
  let app = express();
  const router = ctx.app(api, { express: app });

  router.use(session({ secret }));

  router.post('/sessions', (req, res) => {
    if (req.session?.role != null) {
      res.status(400).json({
        status: 'error',
        message: 'Client already has a session',
      });
      return;
    }
    const { role, password } = req.body;
    if (role === 'admin' && password !== process.env.ADMIN_PASSWORD) {
      res.status(403).json({
        status: 'error',
        message: `Forbidden role: ${role}`,
      });
    }
    req.session = { role, runs: [] };
    res.status(200).json({ status: 'ok', role, runs: [] });
  });

  router.get('/sessions/current', (req, res) => {
    if (req.session?.role == null) {
      res.status(404).json({
        status: 'error',
        message: 'No session found',
      });
      return;
    }
    res.status(200).json({
      status: 'ok',
      role: req.session.role,
      runs: req.session.runs.map((r) => ({
        id: r.runId,
        experiment: r.experimentId,
      })),
    });
  });

  router.delete('/sessions/current', (req, res) => {
    if (req.session?.role == null) {
      res.status(404).json({
        status: 'error',
        message: 'No session found',
      });
      return;
    }
    req.session = null;
    res.status(200).json({ status: 'ok' });
  });

  router.post('/experiments/runs', async (req, res, next) => {
    try {
      if (req.session?.role == null) {
        req.session = { role: 'participant', runs: [] };
      }
      if (req.session.runId != null) {
        res.status(403).json({
          status: 'error',
          message: 'Client already has a started run, end it first',
        });
        return;
      }
      let runId = req.body?.id ?? cuid();
      let experimentId = req.body?.experiment ?? 'default';
      const run = {
        ...req.body,
        experimentId,
        runId,
        createdAt: new Date(),
      };
      await store.addRun(run);
      req.session.runs.push({ runId, experimentId });
      res.status(200).json({
        status: 'ok',
        run: runId,
        experiment: experimentId,
        links: {
          logs: `/experiments/${experimentId}/runs/${runId}/logs`,
          run: `/experiments/${experimentId}/runs/${runId}`,
        },
      });
    } catch (e) {
      if (e instanceof SqliteError && e.code === 'SQLITE_CONSTRAINT') {
        res.status(400).json({
          status: 'error',
          message:
            'Could not add run, probably because a run with that ID already exists for this experiment',
        });
        return;
      }
      next(e);
    }
  });

  router.put('/experiments/:experiment/runs/:run', async (req, res, next) => {
    let { experiment: experimentId, run: runId } = req.params;
    experimentId = String(experimentId);
    runId = String(runId);
    try {
      if (
        req.session?.runs?.find(
          (r) => r.runId === runId && r.experimentId === experimentId
        ) == null
      ) {
        res.status(403).json({
          status: 'error',
          message: `Client does not have permission to update run "${runId}" of experiment "${experimentId}"`,
        });
        return;
      }
      let run = await store.getRun(experimentId, runId);
      if (run == null) {
        throw new Error(`Session run not found: ${runId}`);
      }

      // At the moment, the only supported put operation is
      // { "status": "completed" | "canceled" }, so there is nothing more to
      // check here, zodios does it for us already.
      if (run.status != 'running') {
        res.status(400).json({
          status: 'error',
          message: 'Run already ended',
        });
        return;
      }
      await store.setRunStatus(experimentId, runId, req.body.status);
      req.session.runs = req.session.runs.filter(
        (r) => r.runId !== runId && r.experimentId !== experimentId
      );
      res.status(200).json({ status: 'ok' });
    } catch (e) {
      next(e);
    }
  });

  router.post(
    '/experiments/:experiment/runs/:run/logs',
    async (req, res, next) => {
      let { experiment: experimentId, run: runId } = req.params;
      experimentId = String(experimentId);
      runId = String(runId);
      try {
        if (
          req.session?.runs?.find(
            (r) => r.runId === runId && r.experimentId === experimentId
          ) == null
        ) {
          res.status(403).json({
            status: 'error',
            message: `Client does not have permission to post logs for run "${runId}" of experiment "${experimentId}"`,
          });
          return;
        }
        let sessionRun = await store.getRun(experimentId, runId);
        if (sessionRun == null) {
          // This should not happen in normal use, except if the database is
          // corrupted, or removed.
          throw new Error(`Session run not found: ${runId}`);
        }
        if (sessionRun.status != 'running') {
          res.status(403).json({
            status: 'error',
            message: 'Cannot add logs to an ended run',
          });
          return;
        }
        let logs = 'logs' in req.body ? req.body.logs : [req.body.log];
        await store.addRunLogs(experimentId, runId, logs);
        res.status(200).json({ status: 'ok' });
      } catch (e) {
        next(e);
      }
    }
  );

  router.get('/experiments/:experiment/runs/logs', async (req, res, next) => {
    try {
      if (req.session?.role !== 'admin') {
        res.status(403).json({
          status: 'error',
          message: 'Access restricted.',
        });
        return;
      }
      let { format, type } = req.query;
      let { experiment } = req.params ?? {};
      let filter: LogFilter = {
        experiment: experiment == null ? undefined : String(experiment),
        type: type,
      };
      if (format === 'csv') {
        res.setHeader('Content-Type', 'text/csv');
        await pipeline(csvExportStream(store, filter), res);
      } else {
        res.setHeader('Content-Type', 'application/json');
        await pipeline(jsonExportStream(store, filter), res);
      }
    } catch (e) {
      next(e);
    }
  });

  router.use('*', (req, res) => {
    res.status(404).json({
      status: 'error',
      message: 'Not found',
    });
  });

  app.use((error: unknown, req: Request, res: Response, next: NextFunction) => {
    if (res.headersSent) {
      next(error);
    } else if (error instanceof Error) {
      res.status(500).json({
        status: 'error',
        message: error.message,
      });
    } else {
      log.error('error', error);
      next(error);
    }
  });

  return app;
}
