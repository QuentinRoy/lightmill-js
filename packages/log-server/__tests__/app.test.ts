import request from 'supertest';
import { afterEach, describe, beforeEach, it, vi, expect, Mock } from 'vitest';
import { Log, Store, StoreError } from '../src/store.js';
import { createLogServer } from '../src/app.js';

type MockStore = {
  [K in keyof Store]: Store[K] extends (...args: infer A) => infer R
    ? Mock<A, R>
    : Store[K];
};

function MockStore(): MockStore {
  return {
    addRun: vi.fn(
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      async (...args) => {
        return { runId: 'addRun:runId', experimentId: 'addRun:experimentId' };
      },
    ),
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    getRun: vi.fn(async (...args) => {
      return {
        runId: 'getRun:runId',
        experimentId: 'getRun:experimentId',
        createdAt: vi.getMockedSystemTime(),
        status: 'running' as const,
      };
    }),
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    setRunStatus: vi.fn((...args) => Promise.resolve()),
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    addRunLogs: vi.fn((...args) => Promise.resolve()),
    getLogValueNames: vi.fn(() => Promise.resolve(['mock-col1', 'mock-col2'])),
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    getLogs: vi.fn(async function* (): AsyncGenerator<Log> {
      yield {
        experimentId: 'getLogs:experimentId-1',
        runId: 'getLogs:runId-1',
        type: 'getLogs:type-1',
        createdAt: vi.getMockedSystemTime()!,
        clientDate: new Date(0),
        values: {
          'mock-col1': 'log1-mock-value1',
          'mock-col2': 'log1-mock-value2',
        },
      };
      yield {
        experimentId: 'getLogs:experimentId-2',
        runId: 'getLogs:runId-2',
        type: 'getLogs:type-2',
        createdAt: vi.getMockedSystemTime()!,
        clientDate: new Date(0),
        values: {
          'mock-col1': 'log2-mock-value1',
          'mock-col2': 'log2-mock-value2',
          'mock-col3': 'log2-mock-value3',
        },
      };
    }),
  };
}

afterEach(() => {
  vi.resetAllMocks();
});

describe('/sessions', () => {
  let api: request.SuperTest<request.Test>;
  beforeEach(() => {
    let app = createLogServer({
      store: MockStore(),
      secret: 'secret',
      secureCookies: false,
    });
    api = request.agent(app);
  });

  describe('post /sessions', () => {
    it('should create a session', async () => {
      await api.post('/sessions').send({ role: 'participant' }).expect(201, {
        role: 'participant',
        runs: [],
        status: 'ok',
      });
    });

    it('should refuse to create a session for an unknown role', async () => {
      await api.post('/sessions').send({ role: 'fake' }).expect(400);
    });

    it('should accept the creation of an admin session if there is no admin password set on the server', async () => {
      await api.post('/sessions').send({ role: 'admin' }).expect(201, {
        role: 'admin',
        runs: [],
        status: 'ok',
      });
    });

    it('should accept the creation of an admin role if the provided password is correct', async () => {
      let app = createLogServer({
        store: MockStore(),
        secret: 'secret',
        adminPassword: 'admin password',
      });
      let api = request(app);
      await api
        .post('/sessions')
        .send({ role: 'admin', password: 'admin password' })
        .expect(201, {
          role: 'admin',
          runs: [],
          status: 'ok',
        });
    });

    it('should refuse the creation of an admin role if the provided password is incorrect', async () => {
      let app = createLogServer({
        store: MockStore(),
        secret: 'secret',
        adminPassword: 'admin password',
      });
      let req = request(app);
      await req
        .post('/sessions')
        .send({ role: 'admin', password: 'not the admin password' })
        .expect(403, {
          message: 'Forbidden role: admin',
          status: 'error',
        });
    });
  });

  describe('get /sessions/current', () => {
    it('should return an error if the session does not exists', async () => {
      await api.get('/sessions/current').expect(404, {
        message: 'No session found',
        status: 'error',
      });
    });

    it('should return the current session if it exists', async () => {
      await api.post('/sessions').send({ role: 'participant' }).expect(201);
      await api.get('/sessions/current').expect(200, {
        role: 'participant',
        runs: [],
        status: 'ok',
      });
    });
  });

  describe('delete /sessions/current', () => {
    it('should return an error if the session does not exists', async () => {
      await api.delete('/sessions/current').expect(404, {
        message: 'No session found',
        status: 'error',
      });
    });

    it('should delete the current session if it exists', async () => {
      await api.post('/sessions').send({ role: 'participant' }).expect(201);
      await api.delete('/sessions/current').expect(200, { status: 'ok' });
      await api.get('/sessions/current').expect(404);
    });
  });
});

describe('/experiments/runs', () => {
  let store: MockStore;
  let api: request.SuperTest<request.Test>;
  beforeEach(async () => {
    vi.useFakeTimers({ toFake: ['Date'] });
    vi.setSystemTime(1234567890);
    store = MockStore();
    let app = createLogServer({
      store,
      secret: 'secret',
      secureCookies: false,
    });
    api = request.agent(app);
    await api.post('/sessions').send({ role: 'participant' });
  });
  afterEach(() => {
    vi.useRealTimers();
  });

  describe('post /experiments/runs', () => {
    it('should create a run', async () => {
      await api
        .post('/experiments/runs')
        .send({ experiment: 'exp-id', id: 'run-id' })
        .expect(201, {
          experiment: 'exp-id',
          links: {
            logs: '/experiments/exp-id/runs/run-id/logs',
            run: '/experiments/exp-id/runs/run-id',
          },
          run: 'run-id',
          status: 'ok',
        });
      expect(store.addRun).toHaveBeenCalledWith({
        experimentId: 'exp-id',
        runId: 'run-id',
        createdAt: vi.getMockedSystemTime(),
      });
    });

    it('should use the default experiment if the experiment is not named', async () => {
      await api
        .post('/experiments/runs')
        .send({ id: 'run' })
        .expect(201, {
          status: 'ok',
          experiment: 'default',
          run: 'run',
          links: {
            logs: '/experiments/default/runs/run/logs',
            run: '/experiments/default/runs/run',
          },
        });
      expect(store.addRun).toHaveBeenCalledWith({
        experimentId: 'default',
        runId: 'run',
        createdAt: vi.getMockedSystemTime(),
      });
    });

    it('should generate a unique run id if the run id is not provided', async () => {
      let resp = await api
        .post('/experiments/runs')
        .send({ experiment: 'exp' })
        .expect(201);
      let runId = resp.body.run;
      expect(runId).toBeDefined();
      expect(resp.body).toEqual({
        status: 'ok',
        experiment: 'exp',
        run: runId,
        links: {
          logs: `/experiments/exp/runs/${runId}/logs`,
          run: `/experiments/exp/runs/${runId}`,
        },
      });
      expect(store.addRun).toHaveBeenCalledWith({
        experimentId: 'exp',
        runId,
        createdAt: vi.getMockedSystemTime(),
      });
    });

    it('should add the run to the session', async () => {
      await api
        .post('/experiments/runs')
        .send({ experiment: 'exp-id', id: 'run-id' })
        .expect(201);
      await api.get('/sessions/current').expect(200, {
        role: 'participant',
        runs: [{ id: 'run-id', experiment: 'exp-id' }],
        status: 'ok',
      });
    });

    it('should refuse to create a run if the participant already has one running', async () => {
      await api
        .post('/experiments/runs')
        .send({ experiment: 'exp-id', id: 'run1' })
        .expect(201);
      await api
        .post('/experiments/runs')
        .send({ experiment: 'exp-id', id: 'run2' })
        .expect(403, {
          message: 'Client already has a started run, end it first',
          status: 'error',
        });
    });

    it('should refuse to create a run if the run id already exists', async () => {
      store.addRun.mockImplementation(async ({ runId }) => {
        throw new StoreError(`run "${runId}" already exists`, 'RUN_EXISTS');
      });
      await api
        .post('/experiments/runs')
        .send({ experiment: 'exp-id', id: 'run-id' })
        .expect(400, {
          status: 'error',
          message: 'run "run-id" already exists',
        });
    });
  });

  describe('put /experiments/:experiment/runs/:run', () => {
    it('should return an error if the client does not have access to any run', async () => {
      await api
        .put('/experiments/exp/runs/not-my-run')
        .send({ status: 'completed' })
        .expect(403, {
          status: 'error',
          message: `Client does not have permission to update run "not-my-run" of experiment "exp"`,
        });
    });
    it('should return an error if the client does not have access to this particular run', async () => {
      await api
        .post('/experiments/runs')
        .send({ experiment: 'exp-id', id: 'my-run' })
        .expect(201);
      await api
        .put('/experiments/exp/runs/not-my-run')
        .send({ status: 'completed' })
        .expect(403, {
          status: 'error',
          message: `Client does not have permission to update run "not-my-run" of experiment "exp"`,
        });
    });
    it('should complete a running run if argument is "completed"', async () => {
      await api
        .post('/experiments/runs')
        .send({ experiment: 'exp-id', id: 'my-run' })
        .expect(201);
      await api
        .put('/experiments/exp-id/runs/my-run')
        .send({ status: 'completed' })
        .expect(200, { status: 'ok' });
      expect(store.setRunStatus).toHaveBeenCalledWith(
        'exp-id',
        'my-run',
        'completed',
      );
    });
    it('should cancel a running run if argument is "canceled"', async () => {
      await api
        .post('/experiments/runs')
        .send({ experiment: 'exp-id', id: 'my-run' })
        .expect(201);
      await api
        .put('/experiments/exp-id/runs/my-run')
        .send({ status: 'canceled' })
        .expect(200, { status: 'ok' });
      expect(store.setRunStatus).toHaveBeenCalledWith(
        'exp-id',
        'my-run',
        'canceled',
      );
    });
    it('should revoke client access to the run', async () => {
      await api
        .post('/experiments/runs')
        .send({ experiment: 'exp', id: 'my-run' })
        .expect(201);
      await api
        .put('/experiments/exp/runs/my-run')
        .send({ status: 'completed' })
        .expect(200);
      await api.get('/sessions/current').expect(200, {
        role: 'participant',
        runs: [],
        status: 'ok',
      });
      await api
        .put('/experiments/exp/runs/my-run')
        .send({ status: 'canceled' })
        .expect(403, {
          status: 'error',
          message: `Client does not have permission to update run "my-run" of experiment "exp"`,
        });
    });
  });
});
