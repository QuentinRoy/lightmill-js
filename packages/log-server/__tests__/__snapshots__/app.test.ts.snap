// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`logs > get /experiments/:experiment/logs > should be able to filter logs by type using the type query parameter 1`] = `
[
  [
    {
      "experimentName": "exp",
      "runStatus": [
        "-canceled",
      ],
      "type": "log-type",
    },
  ],
]
`;

exports[`logs > get /experiments/:experiment/logs > should return logs as csv if csv is the first supported format in the Accept header 1`] = `
[
  [
    {
      "experimentName": "exp",
      "runStatus": "-canceled",
    },
  ],
]
`;

exports[`logs > get /experiments/:experiment/logs > should return logs as csv if csv is the first supported format in the Accept header 2`] = `
"type,run_name,run_status,mock_col1,mock_col2,mock_col3
getLogs:type-1,getLogs:runName-1,running,log1-mock-value1,log1-mock-value2,
getLogs:type-2,getLogs:runName-2,completed,log2-mock-value1,log2-mock-value2,log2-mock-value3
"
`;

exports[`logs > get /experiments/:experiment/logs > should return logs as json by default 1`] = `
[
  [
    {
      "experimentName": "exp",
      "runStatus": [
        "-canceled",
      ],
    },
  ],
]
`;

exports[`logs > get /experiments/:experiment/logs > should return logs as json by default 2`] = `
[
  {
    "experimentName": "getLogs:experimentName-1",
    "runName": "getLogs:runName-1",
    "runStatus": "running",
    "type": "getLogs:type-1",
    "values": {
      "mock-col1": "log1-mock-value1",
      "mock-col2": "log1-mock-value2",
    },
  },
  {
    "experimentName": "getLogs:experimentName-2",
    "runName": "getLogs:runName-2",
    "runStatus": "completed",
    "type": "getLogs:type-2",
    "values": {
      "mock-col1": "log2-mock-value1",
      "mock-col2": "log2-mock-value2",
      "mock-col3": "log2-mock-value3",
    },
  },
]
`;

exports[`logs > get /experiments/:experiment/logs > should return logs as json if json is the first supported format in the Accept header 1`] = `
[
  [
    {
      "experimentName": "exp",
      "runStatus": [
        "-canceled",
      ],
    },
  ],
]
`;

exports[`logs > get /experiments/:experiment/logs > should return logs as json if json is the first supported format in the Accept header 2`] = `
[
  {
    "experimentName": "getLogs:experimentName-1",
    "runName": "getLogs:runName-1",
    "runStatus": "running",
    "type": "getLogs:type-1",
    "values": {
      "mock-col1": "log1-mock-value1",
      "mock-col2": "log1-mock-value2",
    },
  },
  {
    "experimentName": "getLogs:experimentName-2",
    "runName": "getLogs:runName-2",
    "runStatus": "completed",
    "type": "getLogs:type-2",
    "values": {
      "mock-col1": "log2-mock-value1",
      "mock-col2": "log2-mock-value2",
      "mock-col3": "log2-mock-value3",
    },
  },
]
`;

exports[`logs > get /experiments/:experiment/logs > should return logs as json if the Accept header is not supported 1`] = `
[
  [
    {
      "experimentName": "exp",
      "runStatus": [
        "-canceled",
      ],
    },
  ],
]
`;

exports[`logs > get /experiments/:experiment/logs > should return logs as json if the Accept header is not supported 2`] = `
[
  {
    "experimentName": "getLogs:experimentName-1",
    "runName": "getLogs:runName-1",
    "runStatus": "running",
    "type": "getLogs:type-1",
    "values": {
      "mock-col1": "log1-mock-value1",
      "mock-col2": "log1-mock-value2",
    },
  },
  {
    "experimentName": "getLogs:experimentName-2",
    "runName": "getLogs:runName-2",
    "runStatus": "completed",
    "type": "getLogs:type-2",
    "values": {
      "mock-col1": "log2-mock-value1",
      "mock-col2": "log2-mock-value2",
      "mock-col3": "log2-mock-value3",
    },
  },
]
`;
